util/types/json/functions.go

```diff
@@ -165,3 +165,125 @@ func extract(j JSON, pathExpr PathExpression) (ret []JSON) {
 	}
 	return
 }
+
+// autoWrapAsArray wraps input JSON into an array if needed.
+func autoWrapAsArray(j JSON, hintLength int) JSON {
+	jnew := CreateJSON(nil)
+	jnew.typeCode = typeCodeArray
+	jnew.array = make([]JSON, 0, hintLength)
+	jnew.array = append(jnew.array, j)
+	return jnew
+}
+
+// Merge merges suffixes into j according the following rules:
+// 1) adjacent arrays are merged to a single array;
+// 2) adjacent object are merged to a single object;
+// 3) a scalar value is autowrapped as an array before merge;
+// 4) an adjacent array and object are merged by autowrapping the object as an array.
+func (j JSON) Merge(suffixes []JSON) JSON {
```

Comment:
How about this? I think it would make the logic simpler.

```
func (j JSON) Merge(suffixes []JSON) JSON {
	for s := range suffixes {
		j = mergeTwo(j, s)
	}
	return j
}

func mergeTwo(a, b JSON) JSON {
	if a.typeCode == typeCodeObject && b.typeCode == typeCodeObject {
		// special case when both are object
		return
	}

	if b.typeCode == typeCodeArray {
		a, b = b, a // swap array to first
	}
	if a.typeCode != typeCodeArray {
		a = autoWrapAsArray(a, len(suffixes)+1)
	}
   // after this, a must be array, so only consider b

	switch b.typeCode {
	case typeCodeArray:
		// merge two array
	default:
		// append b to a
	}
	return a
}
```